File: src/main/java/tmi/recipe/parser/HeatGeneratorParser.java
package tmi.recipe.parser;

import arc.struct.Seq;
import mindustry.content.Blocks;
import mindustry.world.Block;
import mindustry.world.blocks.power.ConsumeGenerator;
import mindustry.world.blocks.power.HeaterGenerator;
import mindustry.world.meta.StatUnit;
import tmi.recipe.Recipe;
import tmi.recipe.RecipeParser;
import tmi.recipe.RecipeType;
import tmi.recipe.types.HeatMark;
import tmi.recipe.types.PowerMark;

public class HeatGeneratorParser extends ConsumerParser<HeaterGenerator>{
  {
    excludes.add(GeneratorParser.class);
    excludes.add(ConsGeneratorParser.class);
  }

  @Override
  public boolean isTarget(Block content) {
    return content instanceof HeaterGenerator;
  }

  @Override
  public Seq<Recipe> parse(HeaterGenerator content) {
    Recipe res = new Recipe(RecipeType.generator);
    res.setBlock(getWrap(content));
    res.setTime(content.itemDuration);

    registerCons(res, content.consumers);

    res.addProductionPresec(PowerMark.INSTANCE, content.powerProduction);

    if (content.heatOutput > 0){
      res.addProductionRaw(HeatMark.INSTANCE, content.heatOutput).setFloatFormat();
    }

    if (content.outputLiquid != null){
      res.addProductionPresec(getWrap(content.outputLiquid.liquid), content.outputLiquid.amount);
    }

    return Seq.with(res);
  }
}

File: src/main/kotlin/tmi/recipe/parser/HeatGeneratorParser.kt
package tmi.recipe.parser

import arc.struct.Seq
import mindustry.world.Block
import mindustry.world.blocks.power.HeaterGenerator
import tmi.recipe.Recipe
import tmi.recipe.RecipeParser
import tmi.recipe.types.RecipeItemType
import tmi.recipe.RecipeType
import tmi.recipe.types.HeatMark
import tmi.recipe.types.PowerMark

open class HeatGeneratorParser : ConsumerParser<HeaterGenerator>() {
  override val excludes: Seq<Class<out RecipeParser<*>>> = Seq.with(
    GeneratorParser::class.java,
    ConsumeGeneratorParser::class.java
  )

  override fun isTarget(content: Block): Boolean {
    return content is HeaterGenerator
  }

  override fun parse(content: HeaterGenerator): Seq<Recipe> {
    val res = Recipe(
      recipeType = RecipeType.generator,
      ownerBlock = content.getWrap(),
      craftTime = content.itemDuration,
    )

    registerCons(res, *content.consumers)

    res.addProductionPersec(PowerMark, content.powerProduction)
      .setType(RecipeItemType.POWER)

    if (content.heatOutput > 0) {
      res.addProduction(HeatMark, content.heatOutput as Number)
        .setType(RecipeItemType.POWER)
        .floatFormat()
    }

    if (content.outputLiquid != null) {
      res.addProductionPersec(content.outputLiquid.liquid.getWrap(), content.outputLiquid.amount)
        .also { if(content.explodeOnFull) it.setType(RecipeItemType.GARBAGE) }
    }

    return Seq.with(res)
  }
}
